common/src/main/java/com/wynntils/features/combat/AutoAttackFeature.java

```diff
@@ -0,0 +1,122 @@
+
+    @SubscribeEvent
+    public void onChangeCarriedItemEvent(ChangeCarriedItemEvent event) {
+        weaponStatus = WeaponStatus.UNKNOWN;
+    }
+
+    @SubscribeEvent
+    public void onSetSlotEvent(SetSlotEvent.Post event) {
+        if (event.getSlot() == McUtils.inventory().selected) {
+            weaponStatus = WeaponStatus.UNKNOWN;
+        }
+    }
+
+    @SubscribeEvent
+    public void onSwing(ArmSwingEvent event) {
+        if (event.getActionContext() != ArmSwingContext.ATTACK_OR_START_BREAKING_BLOCK) return;
+        if (event.getHand() != InteractionHand.MAIN_HAND) return;
+        if (Models.Character.getClassType() != ClassType.ARCHER) return;
+        lastSelectedSlot = McUtils.inventory().selected;
+        preventWrongCast = McUtils.player().tickCount + SPELL_TIMEOUT_TICKS;
+    }
+
+    @SubscribeEvent
+    public void onUseItem(UseItemEvent event) {
+        if (Models.Character.getClassType() == ClassType.ARCHER) return;
+        lastSelectedSlot = McUtils.inventory().selected;
+        preventWrongCast = McUtils.player().tickCount + SPELL_TIMEOUT_TICKS;
+    }
+
+    @SubscribeEvent
+    public void onSpellCastCompleted(SpellEvent.Completed event) {
+        preventWrongCast = Integer.MIN_VALUE;
+    }
+
+    private boolean isHoldingUsableWeapon() {
+        ItemStack heldItem = McUtils.player().getItemInHand(InteractionHand.MAIN_HAND);
+        if (!ItemUtils.isWeapon(heldItem)) return false;
+        if (ItemUtils.getItemName(heldItem).contains("Unidentified")) return false;
+        for (StyledText lore : LoreUtils.getLore(heldItem)) {
+            if (lore.contains("✖")) return false;
+        }
+        return true;
+    }
+
+    @SubscribeEvent
+    public void onTick(TickEvent event) {
+        if (!Models.WorldState.onWorld()) return;
+
+        LocalPlayer player = McUtils.player();
+        int currentSelectedSlot = McUtils.inventory().selected;
+        if (currentSelectedSlot == lastSelectedSlot) {
+            if (preventWrongCast > player.tickCount) return;
+        } else {
+            lastSelectedSlot = currentSelectedSlot;
+            preventWrongCast = Integer.MIN_VALUE;
+        }
+
+        if (player.tickCount % TICKS_PER_ATTACK != 0) return;
+
+        if (weaponStatus == WeaponStatus.UNKNOWN) {
+            weaponStatus = isHoldingUsableWeapon() ? WeaponStatus.USABLE : WeaponStatus.NOT_USABLE;
+        }
+
+        if (weaponStatus != WeaponStatus.USABLE) return;
+
+        boolean isArcher = Models.Character.getClassType() == ClassType.ARCHER;
+        if (isArcher) {
+            if (!McUtils.options().keyUse.isDown()) return;
+        } else {
+            if (!McUtils.options().keyAttack.isDown()) return;
+        }
+
+        if (isArcher) {
+            SpellDirection.RIGHT.getSendPacketRunnable().run();
+        } else {
+            // SpellDirection.LEFT doesn't do the swing animation
+            player.swing(InteractionHand.MAIN_HAND);
```

Comment:
Why not combine these if's, then you don't need to save the isArcher boolean?

Reply:
Good point. :-)